src/sentry/taskworker/processors/strategy_factory.py

```diff
@@ -86,18 +90,32 @@ def flush_batch(
             return message
 
         def do_upkeep(
-            message: Message[MutableSequence[InflightActivation]],
-        ) -> Message[MutableSequence[InflightActivation]]:
+            message: Message[KafkaPayload],
+        ) -> KafkaPayload:
             self.pending_task_store.handle_processing_deadlines()
             self.pending_task_store.handle_retry_state_tasks()
             self.pending_task_store.handle_deadletter_at()
             self.pending_task_store.handle_failed_tasks()
-
-            return message
-
-        upkeep_step = RunTask(
-            function=do_upkeep,
-            # TODO ideally we commit offsets for completed work
+            return message.payload
+
+        def limit_tasks(
+            message: Message[KafkaPayload],
+        ) -> KafkaPayload:
+            count = self.pending_task_store.count_pending_task()
+            while count >= self.max_inflight_activation_in_store:
+                # The number of pending inflight activations in the store exceeds the limit.
+                # Wait for workers to complete tasks before adding the next offset to the queue.
+                logger.info(
+                    "Number of inflight activation: %s exceeds the limit: %s. Retrying in 3 seconds",
+                    count,
+                    self.max_inflight_activation_in_store,
+                )
+                time.sleep(3)  # TODO: make this a parameter
```

Comment:
We should emit backpressure by raising a `MessageRejected` error.

It's not recommended to sleep in submit.
https://getsentry.github.io/arroyo/backpressure.html#backpressure

Reply:
Good callout, and kudos to whoever wrote the documentation for it. Let me know how the changes look.